atmel-samd/tick.c

```diff
@@ -29,18 +44,21 @@ void timer_tick(const struct timer_task *const timer_task) {
 }
 
 void tick_init() {
-    #ifdef SAMD21
-    _pm_enable_bus_clock(PM_BUS_APBC, TC5);
-    #endif
-    // _gclk_enable_channel(TC5_GCLK_ID, GCLK_SOURCE_DFLL48M);
-
-    // timer_init(&ms_timer, TC5, _tc_get_timer());
+    SysTick->LOAD = ((CONF_CPU_FREQUENCY / 1000 - 1) << SysTick_LOAD_RELOAD_Pos);
+    NVIC_EnableIRQ(SysTick_IRQn);
+    SysTick->CTRL = (1 << SysTick_CTRL_ENABLE_Pos) |
+                    (1 << SysTick_CTRL_TICKINT_Pos)|
+                    (1 << SysTick_CTRL_CLKSOURCE_Pos);
+}
 
-    // timer_set_clock_cycles_per_tick(&ms_timer, 48000000 / 1000 - 1);
-    // task.cb       = timer_tick;
-    // task.interval = 1;
-    // task.mode     = TIMER_TASK_REPEAT;
-    // timer_add_task(&ms_timer, &task);
-    //
-    // timer_start(&ms_timer);
+void tick_delay(uint32_t us) {
+    uint32_t fractional = SysTick->VAL;
+    uint64_t start_ms = ticks_ms;
+    while (us > 1000) {
+        while (ticks_ms == start_ms) {}
+        us -= fractional;
```

Comment:
Based on my Googling, `SysTick` units are dependent on clock frequency, right? The SysTick timer is decremented on each clock cycle. So for instance at 48MHz, one tick is 1sec/48e6. So are you mixing microsoeconds (`us`) with SysTicks units?

Reply:
Yes! You are right.